### ISSUE TITLE
[Bug]:  Relative navigation issue while navigating from "index" route: ".." doesn't works, I must use "../.." to go back

### ISSUE THREAD

**Issue by Gryndii** [remix-run/react-router#8350]
### What version of React Router are you using?

6.0.2

### Steps to Reproduce

```
<Some other routes>
....
<Route path='/pages'>
    <Route index element={<Pages />} />

    <Route path=':pageId' element={<Item />} />              
  </Route>
  
```
  
  Navigation via '..' from Pages with "index" doesn't works but it is ok for Item
  I should use '../..' for the same result as '..'

### Expected Behavior

Should navigate level up via '..' from index route
And it would be nice to have option with nested routes that is used only to create url hierarchy for relative navigation usage, but without children and parrents ui rendering

### Actual Behavior

doesnt navigate from index page via '..'

**Comment by maxandriani** [comment]
I'm struggling w/ the same issue, but I think the root cause is because the "base" url should contains a trailing slash, otherwise the history dependency will compute the relative path wrong.

The problems appears to be in the `resolve-pathname.js -> resolve-pathname -> history -> react-router` dependency.

In the line 30, the algorithm just consider everything in the same level as a filename string w/o check if it is really a filename string (file.ext). 

![image](https://user-images.githubusercontent.com/7096439/143437143-f9923be7-3bbb-40ab-ac41-4c84a484f60f.png)

**Comment by mrlika** [comment]
I think it is because `..` means navigate one `<Route>` back and not one directory back. Consider this example:

```jsx
function NavigateBack() {
  return <Navigate to='..' />;
}

<Route path='/test'>
  <Route path=':a/:b/:c' element={<NavigateBack />} />
  <Route path=':a/:b' element={<NavigateBack />} />
  <Route path=':a' element={<NavigateBack />} />
</Route>
```

`/test/a/b/c` `/test/a/b`, and `/test/a` all navigate to `/test` that is really weird. I can't find this in documentation.

**Comment by gowthamvbhat** [comment]
> Both `/test/a/b/c` and `/test/a` are redirected to `/test` that is really weird. I can't find this in documentation.

@mrlika This is discussed here - [upgrading to v6](https://reactrouter.com/docs/en/v6/upgrading/v5#note-on-link-to-values).

**Comment by mrlika** [comment]
@gowthamvbhat, thanks for pointing out that. As for me, it is better to write `unlike <a href> that goes up by path segments`.

**Comment by mrlika** [comment]
The workaround to navigate up in path segments is to use `resolvePath` like:

```jsx
const navigate = useNavigate();
navigate(resolvePath('..', window.location.pathname));

// or

<Link to={resolvePath('..', window.location.pathname)} />

// or

<Navigate to={resolvePath('..', window.location.pathname)} />
```

**Comment by Huulivoide** [comment]
It seems that this is as designed
They have tests in place ensuring that `to=".."` navigates back to the same page in index route.

I guess that technically this is "correct" if the documentation says that .. traverses up by one "Route".

Tough I feel like the index routes should maybe be a special case to the rule.

**Comment by shulcsm** [comment]
This is REALLY annoying with index routes and used to work in beta.

Imagine scenario
```jsx
<Route path=":id/">
  <Route index element={<Link to="../../sibling" />} />
  ....
</Route>
```

**Comment by Romcol** [comment]
I have also experienced the bug. This is also a breaking change coming from reach router. It should be fixed or specified in the doc :)

**Comment by JaffParker** [comment]
I experience this problem and it makes kind of no sense how these work. I actually expect it to go up a route in the tree as opposed to just removing one segment (although in this particular instance it's the same). Here are my routes:

```tsx
<Route path="/settings">
  <Route index element={<SettingsRoute />} />
  <Route
    path="account"
    element={<SettingsRoute page="account" />}
  />
  <Route path="team">
    <Route index element={<SettingsRoute page="team" />} />
    <Route
      path="create"
      element={<CreateUserRoute />}
    />
    <Route path=":id">
      <Route index element={<EditUserRoute />} />
      <Route
        path="history"
        element={<UserActivityHistoryRoute />}
      />
    </Route>
  </Route>
</Route>
```

Here's the list of routes I have with the desired `to=".."` behavior:

```
/settings -> nothing, it's the root
/settings/account -> go back to /settings, works
/settings/team -> go back to /settings, does not work since the index' parent route is path="team", which is silly
/settings/team/create -> go back to /settings/team, works
/settings/team/:id -> go back to /settings/team, does not work since the index' parent route is path=":id", again silly
/settings/team/:id/history -> go back to /settings/team/:id, works
```

So the silly behavior comes from how the `index` routes are handled - they're treated as children of the parent route instead of being treated as the parent route itself.

**Basically what I'm saying is, I think `to=".."` inside index routes should be treated as `to="../.."` for proper predictability.** Otherwise, there's no workaround for this right now since I can't just render the element in the parent route, since that'll render it for all child routes; and I can't move everything a level up since that'll break navigation for the rest of the child routes.

**Comment by ryanflorence** [comment]
Whew ... this is a tough one but I get it.

The question is:

1. do we remain consistent that `..` means "up the route hierarchy"
2. or do we add special cases for index and pathless layout routes?

I think everybody is right that (2) is probably better. So the explanation would be

> `".."` means up the route hierarchy, ignoring routes without paths

Seems easy enough to explain and both index and pathless layout routes don't have any paths.

> So the silly behavior...

What's the point of life if we can't be silly sometimes?

**Comment by ryanflorence** [comment]
Trouble is it's a breaking change technically :\ Gonna see if there's a way to make it opt-in.

**Comment by ryanflorence** [comment]
@kiliman I wish I had a link or something to send to you, but we labored over that behavior for months (literally) with dozens of scenarios and are confident traversing the route hierarchy is preferred.

The only question here is do we skip route's that don't add any path segments.

**Comment by brophdawg11** [comment]
This has been fixed in #8697 / #8954.  Routes that do not provide any path should not contribute to relative pathing logic - should be released in 6.4.0

**Comment by brophdawg11** [comment]
This is fixed in the above mentioned PRs - should be included in the next `6.4.0` prerelease 👍

**Comment by kirkobyte** [comment]
Is there any way/workaround to receive the preivous behaviour? I've got a use-case where the previous behaviour was really beneficial around optional props.

In V5 we had a route with an optional parameter (`options/:option?`), which we now render like this:
```tsx
<Route path="options">
  <Route index element={<ListPage />}/>
  <Route path=":option" element={<ListPage/>}/>
 </Route>
```

We were hoping that in `ListPage` we could define our routes with a simple `<Link to={`../<option id>`}>Option</Link>`. But since `index` routes are ignored, the routing ends up leaving `options`.

**Comment by brophdawg11** [comment]
In your case I would probably just use the presence of the `option` param to drive the `to`:

```jsx
let params = useParams();
let to = params.option ? ".." : ".";
...
<Link to={to}>Option</Link>
```

We have been chatting a bit about optional params recently and I think they'll be on the roadmap once we finish getting Remix on top of React Router 6.4, so keep an eye on https://github.com/remix-run/react-router/issues/8381 as we'll post any updates over there if/when we move forward on that front.

**Comment by andi0b** [comment]
@kirkobyte, @brophdawg11: I stumbled into exactly the same issue. I have a module and all my components link relative from there. But once I use an index route, all links break. Which means index routes can't be really used anymore in a meaningful way, because they break relative links without any way of fixing it (there isn't even an opt-in switch on the Route). This new "fixed" behavior is highly illogical for me.

This "new feature" already cost me hours of my productivity, because it's really hard to figure out what "parent route" really means. Route should have a flag like `transparent`, or `canBeParent` which specifies if it should participate in relative routing.

```jsx
<Route path="module">
  <Route index element={<FeatureA/>}/> {/* which breaks all the links, no way to solve it, except handling it inside feature a */}
  <Route path="feature-a" element={<FeatureA/>}/>
  <Route path="feature-b" element={<FeatureB/>}/>
  {...}
 </Route>
```

ugly workaround:
```jsx
<Route path="module">
  <Route index element={<Naviage to='./feature-a' />}/>
  <Route path="feature-a" element={<FeatureA/>}/>
  <Route path="feature-b" element={<FeatureB/>}/>
  {...}
 </Route>
```
